README.md
# pyctd
Python tools to read, process and plot several different conductivity, temperature and depth (CTD) formats including Seabird and Sea &amp; Sun Technology files


pyctd/CHANGELOG.rst
0.4.2:
        - some bugfixes (i.e. crash geojson)
	- improved searching capability (lon,lat, start, stop)
0.4.1:
        - outsourced pysst
	- load/save functionality improved (yaml, geojson for qgis)
	- added transects/campaigns and research vessels
	- infrastructure improved to work as a qgis plugin
0.4.0:
        - outsourced pycnv
	- Added tabs and station table
0.3.9:
        - Moved the whole repo to pyctd, to have a combined too for CTD datasets, versioning and changelog will be kept
0.3.8:
        - added first draft of gui
0.3.7:
        - first version with a rudimentary plot routine
	- changed data from recarray to dictionary
0.3.6:
        - cdata has a copy of the 'p' field for convenience
	- oxygen is now correctly read in as standard name
	- oxygen saved in ml/l is converted to umol/l and saved in cdata
0.3.5:
        - Custom header functionality added
0.3.4:
        - Added requirements to setup.py
        - changed CHANGELOG to CHANGELOG.rst

0.3.3:
        - Changed pycnv.date to start_time field
	- Timezone is now added to date fields
0.3.2:
	- oxygen conversion routines added
0.3.1:
	- added pycnv.regions_baltic for testing of CTD cast is in the Baltic
	- added a conversion of IPTS-68 to T90 and S7m to mS/cm
	- added pycnv.get_all_valid_files()
	- added pycnv.get_stations()
	- added pycnv/test/plot_baltic_test_regions.py
	- added pycnv/test/make_netcdf.py
	- several small stability changes


pyctd/sst/pymrd.py
import datetime
from pytz import timezone
import math
import numpy
import logging
import sys
import argparse
import pkg_resources
import yaml
import pylab as pl
import os
import hashlib

standard_name_file = pkg_resources.resource_filename('pyctd', 'rules/standard_names.yaml')

# Get the version
version_file = pkg_resources.resource_filename('pyctd','VERSION')

with open(version_file) as version_f:
   version = version_f.read().strip()

# Setup logging module
logging.basicConfig(stream=sys.stderr, level=logging.INFO)
logger = logging.getLogger('pymrd')


class pymrd():
    """A object to read and process Sea & Sun Technology MRD files
    generated by the Microstructure profiler MSS.

    """ 
    def __init__(self,filename, only_metadata = False,verbosity =
                 logging.INFO, naming_rules = standard_name_file, baltic=None, calc_sha1=True):
        logger.setLevel(verbosity)
        logger.info(' Opening file: ' + filename)
        self.filename = filename
        self.file_type = ''
        self.channels = []
        self.data = None
        self.date = None        
        self.lon = numpy.NaN
        self.lat = numpy.NaN

        # Plotting variables
        self.figures = []
        self.axes    = []        
        # Opening file for reading
        try:
            self.f = open(self.filename,'rb')
        except Exception as e:
            logger.critical('Could not open file:' + self.filename + '( ' + str(e) + ' )')
            self.valid_mrd = False
            return

        # Opening file for reading and calculating sha1
        try:
            # Calculate a md5 hash
            if(calc_sha1):
               BLOCKSIZE = 65536
               hasher = hashlib.sha1()
               with open(self.filename, 'rb') as afile:
                   buf = afile.read(BLOCKSIZE)
                   while len(buf) > 0:
                      hasher.update(buf)
                      buf = afile.read(BLOCKSIZE)

               self.sha1 = hasher.hexdigest()
               afile.close()               
            else:
               self.sha1 = None
        except:
            logger.critical('Could not open file:' + self.filename)
            self.valid_mrd = False
            return
        
        self.read_MRD(pos_time_only = only_metadata) # Reading the binary 
        self._parse_header()

    def _parse_header(self):
        """ Parsing the header of the MRD file
        """
        header = self.header

        ind_ship = header.find('Ship    :   ') + len('Ship    :   ')
        ship = header[ind_ship:ind_ship+8]
        ship = ship.rstrip('_')

        ind_cruise = header.find('Cruise:') + len('Cruise:')
        cruise = header[ind_cruise+1:ind_cruise+18]
        cruise = cruise.rstrip('_')
        hs = header.split('\\r')
        sensor_str = []
        mss = None
        for i in range(17,len(hs)-1):
            sensor_str.append(hs[i])
            #if(hs[i].find('107') >=0):
            if(hs[i].upper().find('COUNT') >=0):
                mss = hs[i].split(' ')[1]

        if mss is not None:
            self.cruise = cruise
            self.ship   = ship
            self.mss    = mss
        else:
            logger.warning('Could not find a MSS')

        
    def get_info_dict(self):
        """ Returns a dictionary with the essential information
        """
        info_dict = {}
        info_dict['lon'] = self.lon
        info_dict['lat'] = self.lat
        info_dict['date'] = self.date
        info_dict['file'] = self.filename
        info_dict['sha1'] = self.sha1
        info_dict['type'] = 'MRD'
        return info_dict
    

    def read_MRD(self,pos_time_only = False):
        """ Getting metadata from an SST MRD file
        Args:
           pos_time_only: Read as much information to get date and position
        """
        
        f = self.f

        n = 0
        np = 0 # Number of packets
        end_of_header = 0
        end_of_header_tmp = 0
        IN_HEADER = True
        header    = b''
        HAVE_TIME = False
        HAVE_POS  = False


        while True:
            if HAVE_TIME & HAVE_POS & pos_time_only:
                break

            if(IN_HEADER):    
                b = f.read(1)
                n += 1        
            else:
                b = f.read(17)
                n += 17
                np += 1 # Number of 17 bytes long data packets (packet type + 8 words (little endian)

            if b != b"":
                # We are in the data body        
                if(IN_HEADER == False):
                    bword = []
                    for i in range(1,16,2): # Reading 16 bytes and treat them as words
                        bword.append(int.from_bytes(b[i:i+2],byteorder='little'))
                    if(b[0] == 1): # Time packet
                        if(HAVE_TIME == False):
                            year = int(bword[0])
                            year = bword[0]
                            month = bword[1]
                            day = bword[2]
                            hour = bword[4]
                            minute = bword[5]
                            second = bword[6]
                            date = datetime.datetime(year,month,day,hour,minute,second,tzinfo=timezone('UTC'))
                            self.date = date
                            HAVE_TIME = True

                    if(b[0] == 3): # Position packet
                        if(HAVE_POS == False):
                            # Latitude
                            latint    = int.from_bytes(b[1:3],byteorder='little')
                            latsign   = (latint & 0x8000) >> 15
                            if(latsign == 0):
                                latsign = -1
                            lat       = latint & 0x1FFF
                            latdeg    = (lat - lat%100)/100
                            latmindec = int.from_bytes(b[3:5],byteorder='little')
                            if(abs(latmindec) > 0):                            
                                digits    = int(math.log10(latmindec)) + 1
                            else:
                                digits    = 0
                                
                            latmin    = lat%100 + latmindec / 10**digits
                            latdec    = latsign * (latdeg + latmin/60)
                            # Longitude
                            lonint    = int.from_bytes(b[5:7],byteorder='little')
                            lonsign   = (lonint & 0x8000) >> 15
                            if(lonsign == 0):
                                lonsign = -1                
                            lon       = lonint & 0x1FFF
                            londeg    = (lon - lon%100)/100
                            lonmindec = int.from_bytes(b[7:9],byteorder='little')
                            if(abs(lonmindec) > 0):
                                digits    = int(math.log10(lonmindec)) + 1
                            else:
                                digits    = 0
                                
                            lonmin    = lon%100 + lonmindec / 10**digits
                            londec    = lonsign * (londeg + lonmin/60 )
                            # Daytime
                            tmp = (bword[6]) + (bword[4]&0x000F) * 100000
                            hour = int((tmp - tmp%10000)/10000)
                            tmp2 = tmp%10000
                            minute = int((tmp2 - tmp2%100)/100)
                            second = tmp2%100
                            self.lon = londec
                            self.lat = latdec                            
                            HAVE_POS = True

                    #if(b[0] == 7):
                    #    print('Sensors',b)
                    #    input('RE')

                else:
                    header += b
                    if(b == b'\x1A'): # Header is filled up with 0x1A until it is dividable by 17
                        end_of_header = n
                        # Check if dividable by 17
                        if((end_of_header)%17 == 0): # Dividable by 17
                            self.valid_mrd = True                            
                            #print('Found a valid header (dividable by 17)')
                            self.header = str(header)
                            IN_HEADER = False

            else:
                break


    def __str__(self):
        """
        String format
        """
        rstr = ""
        rstr += "pymrd of " + self.filename
        rstr += " at Lat: " + str(self.lat)
        rstr += ", Lon: " + str(self.lon)
        rstr += ", Date: " + datetime.datetime.strftime(self.date,'%Y-%m-%d %H:%M:%S')
        return rstr                    



# Main function
def main():
    sum_help         = 'Gives a csv compatible summary'
    plot_help        = 'Plots the MRD file, list the parameters in a comma separated list, e.g. --plot CT00,pt00,oxy use the arguments "show" to immidiately show the figure (will halt the code until the figure is closed) and/or "save" to save the figure'
    plot_prefix_help = 'The prefix before the filename, standars is "./", this is usefule to define a path and/or a fie prefix, e.g. --plot_prefix figures/ctd_casts_of_important_cruise__'
    var_help         = 'Lists all the available variables within the file, separated between the orignal data within the file (data) and the computed data (cdata)'        
    sumhead_help     = 'Gives the header to the csv compatible summary'
    parser = argparse.ArgumentParser()
    parser.add_argument('--variables', '-va', action='store_true', help=var_help)    
    parser.add_argument('--summary', '-s', action='store_true', help=sum_help)
    parser.add_argument('--summary_header', '-sh', action='store_true', help=sumhead_help)
    #https://stackoverflow.com/questions/13346540/argparse-optional-argument-before-positional-argument    
    parser.add_argument('--plot', '-p', nargs='?', help=plot_help)
    parser.add_argument('--plot_prefix', '-pre', nargs='?', help=plot_prefix_help)    
    parser.add_argument('--verbose', '-v', action='count')
    #parser.add_argument('--version', action='store_true')
    parser.add_argument('--version', action='version', version='%(prog)s ' + version)
    parser.add_argument('filename')    
    args = parser.parse_args()
    
    if(args.verbose == None):
        loglevel = logging.WARNING        
    elif(args.verbose == 1):
        loglevel = logging.INFO        
    elif(args.verbose == 2):
        loglevel = logging.DEBUG        
    else:
        loglevel = logging.INFO


    logger.setLevel(loglevel)


    filename = args.filename

    print_summary = args.summary
    print_summary_header = args.summary_header
    
    if(filename != None):
        mrd = pymrd(filename,verbosity=loglevel)
    else:
        #logger.critical('Need a filename')
        print(parser.print_help())


    if(args.variables):
        summary = mrd.get_variables()
        print(summary)

    if(print_summary_header):
        summary = mrd.get_summary(header=True)
        print(summary)
    if(print_summary):
        summary = mrd.get_summary()
        print(summary)

    #
    # Plot the file
    #
    if(args.plot != None):
        FLAG_SHOW = False
        FLAG_SAVE = False
        variables_plot = []
        for var in args.plot.split(','):
            if(var.upper() == 'SHOW'):
                FLAG_SHOW = True
            elif( var.upper() == 'SAVE'):
                FLAG_SAVE = True
            else:
                variables_plot.append(var)

        if(args.plot_prefix == None):
            plot_prefix = "./"
        else:
            plot_prefix = args.plot_prefix

        mrd.plot(xaxis=variables_plot,show=FLAG_SHOW,save=FLAG_SAVE,fig_prefix=plot_prefix)

            
if __name__ == '__main__':
   main()        
        


pyctd/sst/pymrd_sum_folder.py
from . import pymrd as pymrd
import glob
import fnmatch
import os
import sys
import numpy
import argparse
import logging
import pkg_resources
import yaml
from pytz import timezone
import datetime


# Get the version
version_file = pkg_resources.resource_filename('pyctd','VERSION')

with open(version_file) as version_f:
   version = version_f.read().strip()

try:
    from pyproj import Geod
    FLAG_PYPROJ=True
    g = Geod(ellps='WGS84')
except:
    FLAG_PYPROJ=False

# Setup logging module
logging.basicConfig(stream=sys.stderr, level=logging.WARNING)
logger = logging.getLogger('pymrd_sum_folder')


def get_stations():
    stations_file = pkg_resources.resource_filename('pycnv', 'stations/iow_stations.yaml')
    f_stations = open(stations_file)
    # use safe_load instead load
    stations_yaml = yaml.safe_load(f_stations)
    return stations_yaml['stations']


def get_all_valid_files(DATA_FOLDER, loglevel = logging.INFO, station = None, save_summary = False, status_function = None):
    """
    Args:
       DATA_FOLDER: Either list of data_folder or string of one data_folder
       station: MSS cast has to lie within radius around position, given as a list with longitude [decdeg], latitude [decdeg], radius [m], e.g. [20.0,54.0,5000]
       status_function: A function that is called during reading, the function is called with the current filenumber i, the total number of files nf and the filename f, e.g. function(i,nf,f) 
    Returns:
        Dictionary with data
    """

    if(isinstance(DATA_FOLDER, str)):
        DATA_FOLDER = [DATA_FOLDER]
    if station == None:
        FLAG_DIST = False
    else:
        londist   = station[0]
        latdist   = station[1]
        distdist  = station[2]
        FLAG_DIST = True

    #
    # Loop through all subfolders
    #
    matches = []
    for DATA_P in DATA_FOLDER:
        for root, dirnames, fnames in os.walk(DATA_P):
            #print(root,dirnames,fnames)
            for fname in fnmatch.filter(fnames, '*.mrd'):
                matches.append(os.path.join(root, fname))
                #print(matches[-1])
                if(numpy.mod(len(matches),100) == 0):
                    logger.info('Found ' + str(len(matches)) + ' files')

            for fname in fnmatch.filter(fnames, '*.MRD'):
                matches.append(os.path.join(root, fname))
                if(numpy.mod(len(matches),100) == 0):
                    logger.info('Found ' + str(len(matches)) + ' files')
                #print(matches[-1])        


    logger.info('Found ' + str(len(matches)) + ' mrd files in folder(s):' + str(DATA_FOLDER))
    if(len(matches) == 0):
        return {'files':[],'dates':[],'lon':[],'lat':[],'info_dict':[]}
    save_file       = []
    files_date      = []
    file_names_save = []
    files_lon_save  = []
    files_lat_save  = []        
    files_date_save = []
    files_summary   = []
    files_info_dict = []
    
    if(len(matches) > 0):
        # Write the header of the file
        print('Hallo',matches[0])
        mrd = pymrd(matches[0],verbosity=logging.CRITICAL,only_metadata = True)

        # Loop through all files and make summary
        nf = len(matches)
        for i,f in enumerate(matches):
            logger.info('Reading file ' + str(i) +'/' + str(nf) + ': ' + str(f))
            if(status_function is not None): # At the moment used for gui
                print('Status function')
                status_function(i,nf,f)
                
            mrd = pymrd(f,verbosity=loglevel,only_metadata = True)
            if(mrd.valid_mrd):
                files_date.append(mrd.date)
                files_info_dict.append(mrd.get_info_dict())
                #summary = mrd.get_summary()
                FLAG_GOOD = False
                # Check if we are within a distance
                lon = mrd.lon
                lat = mrd.lat                
                if(FLAG_DIST):
                    if(not(numpy.isnan(lon)) and not(numpy.isnan(lat))):
                        az12,az21,dist = g.inv(lon,lat,londist,latdist)
                        if(dist < distdist):
                            FLAG_GOOD = True

                else:
                    FLAG_GOOD = True

                if(FLAG_GOOD):
                    save_file.append(True)
                    file_names_save.append(f)
                    files_date_save.append(mrd.date)
                    files_lon_save.append(lon)
                    files_lat_save.append(lat)
                    #files_summary.append(summary)
                else:
                    save_file.append(False)

        # Save the with respect to date sorted file
        logger.info('Sorting all files')
        print(files_lon_save)
        # Replace invalid dates with an obviously wrong date to be able to sort them
        for i in range(len(files_date_save)):
            if(files_date_save[i] == None):
                files_date_save[i] = datetime.datetime(1,1,1).replace(tzinfo=timezone('UTC'))

        ind_sort = numpy.argsort(files_date_save)                
        file_names_save_sort = list(numpy.asarray(file_names_save)[ind_sort])
        retdata  = {'files':file_names_save_sort,'dates':list(numpy.asarray(files_date_save)[ind_sort]),'lon':list(numpy.asarray(files_lon_save)[ind_sort]),'lat':list(numpy.asarray(files_lat_save)[ind_sort]),'info_dict':list(numpy.asarray(files_info_dict)[ind_sort])}

        print(retdata)
        if save_summary:
            summary_array = numpy.asarray(files_summary)[ind_sort]
            retdata['summary'] = summary_array
        
        return retdata


pyctd/seabird/pycnv_sum_folder.py
#
#import pycnv
from . import pycnv as pycnv
import glob
import fnmatch
import os
import sys
import numpy
import argparse
import logging
import pkg_resources
import yaml
from pytz import timezone
import datetime


# Get the version
version_file = pkg_resources.resource_filename('pyctd','VERSION')

with open(version_file) as version_f:
   version = version_f.read().strip()

try:
    from pyproj import Geod
    FLAG_PYPROJ=True
    g = Geod(ellps='WGS84')
except:
    FLAG_PYPROJ=False

# Setup logging module
logging.basicConfig(stream=sys.stderr, level=logging.WARNING)
logger = logging.getLogger('pycnv_sum_folder')


def get_stations():
    stations_file = pkg_resources.resource_filename('pycnv', 'stations/iow_stations.yaml')
    f_stations = open(stations_file)
    # use safe_load instead load
    stations_yaml = yaml.safe_load(f_stations)
    return stations_yaml['stations']
    
def get_all_valid_files(DATA_FOLDER, loglevel = logging.INFO, station = None, save_summary = False, status_function = None):
    """
    Args:
       DATA_FOLDER: Either list of data_folder or string of one data_folder
       station: CTD cast has to lie within radius around position, given as a list with longitude [decdeg], latitude [decdeg], radius [m], e.g. [20.0,54.0,5000]
       status_function: A function that is called during reading, the function is called with the current filenumber i, the total number of files nf and the filename f, e.g. function(i,nf,f) 
    Returns:
        Dictionary with data
    """

    if(isinstance(DATA_FOLDER, str)):
        DATA_FOLDER = [DATA_FOLDER]
    if station == None:
        FLAG_DIST = False
    else:
        londist   = station[0]
        latdist   = station[1]
        distdist  = station[2]
        FLAG_DIST = True

    #
    # Loop through all subfolders
    #
    matches = []
    for DATA_P in DATA_FOLDER:
        for root, dirnames, fnames in os.walk(DATA_P):
            #print(root,dirnames,fnames)
            for fname in fnmatch.filter(fnames, '*.cnv'):
                matches.append(os.path.join(root, fname))
                #print(matches[-1])
                if(numpy.mod(len(matches),100) == 0):
                    logger.info('Found ' + str(len(matches)) + ' files')

            for fname in fnmatch.filter(fnames, '*.CNV'):
                matches.append(os.path.join(root, fname))
                if(numpy.mod(len(matches),100) == 0):
                    logger.info('Found ' + str(len(matches)) + ' files')
                #print(matches[-1])        


    logger.info('Found ' + str(len(matches)) + ' cnv files in folder(s):' + str(DATA_FOLDER))
    if(len(matches) == 0):
        return {'files':[],'dates':[],'lon':[],'lat':[],'info_dict':[]}
    save_file       = []
    files_date      = []
    file_names_save = []
    files_lon_save  = []
    files_lat_save  = []        
    files_date_save = []
    files_summary   = []
    files_info_dict = []    
    if(len(matches) > 0):
        # Write the header of the file
        print('Hallo',matches[0])
        cnv = pycnv(matches[0],verbosity=logging.CRITICAL)

        # Loop through all files and make summary
        nf = len(matches)
        for i,f in enumerate(matches):
            logger.info('Parsing file ' + str(i) +'/' + str(nf) + ': ' + str(f))
            if(status_function is not None):
                print('Status function')
                status_function(i,nf,f)
            cnv = pycnv(f,verbosity=loglevel)
            if(cnv.valid_cnv):
                files_date.append(cnv.date)
                summary = cnv.get_summary()
                FLAG_GOOD = False
                # Check if we are within a distance
                lon = cnv.lon
                lat = cnv.lat
                if(FLAG_DIST):
                    if(not(numpy.isnan(lon)) and not(numpy.isnan(lat))):
                        az12,az21,dist = g.inv(lon,lat,londist,latdist)
                        if(dist < distdist):
                            FLAG_GOOD = True

                else:
                    FLAG_GOOD = True

                if(FLAG_GOOD):
                    save_file.append(True)
                    file_names_save.append(f)
                    files_date_save.append(cnv.date)
                    files_lon_save.append(lon)
                    files_lat_save.append(lat)
                    files_summary.append(summary)
                    files_info_dict.append(cnv.get_info_dict()) # This will be the standard for future development
                else:
                    save_file.append(False)

        # Save the with respect to date sorted file
        logger.info('Sorting all files')
        # Replace invalid dates with an obviously wrong date to be able to sort them
        for i in range(len(files_date_save)):
            if(files_date_save[i] == None):
                files_date_save[i] = datetime.datetime(1,1,1).replace(tzinfo=timezone('UTC'))

        ind_sort = numpy.argsort(files_date_save)                
        file_names_save_sort = list(numpy.asarray(file_names_save)[ind_sort])
        retdata  = {'files':file_names_save_sort,'dates':list(numpy.asarray(files_date_save)[ind_sort]),'lon':list(numpy.asarray(files_lon_save)[ind_sort]),'lat':list(numpy.asarray(files_lat_save)[ind_sort]),'info_dict':files_info_dict}

        if save_summary:
            summary_array = numpy.asarray(files_summary)[ind_sort]
            retdata['summary'] = summary_array
        
        return retdata


def main():

    example1 = 'Example (searching in folders fahrten.2011 and fahrten.2012 for stations TF0286 within a radius of 5000m: pycnv_sum_folder -d fahrten.201[12]/ -f tf286.txt --station TF0286 5000'

    desc             = 'A pycnv tool which is recursively searching through the given data folder and searching for cnv files. Found cnv files are parsed and a summary is written to the file given by --filename. '
    desc += example1

    data_help        = 'The data path(es) to be searched'
    file_help        = 'The filename the results are written to'
    dist_help        = 'Only take files which are within a distance to CTD location'
    station_help     = 'Only take files which are within a distance to station'
    stationlist_help = 'Lists all known stations with their names and positions'
    print_help       = 'Prints for each line the summary to stdout'
    verb_help        = 'Add -v to increase verbosity of command'
    parser           = argparse.ArgumentParser(description=desc)


    parser.add_argument('--data_folder', '-d'  , nargs = '*', required=False, help=data_help)
    parser.add_argument('--filename', '-f'   , default = None,help=file_help)
    parser.add_argument('--distance', '-dist', nargs=3,metavar=('lon [dec deg]','lat [dec deg]','distance [m]'), help=dist_help)
    parser.add_argument('--station', '-s'    , nargs=2,metavar=('Station name','distance [m]'), help=station_help)
    parser.add_argument('--list_stations'    , '-ls', action='store_true', help=stationlist_help)
    parser.add_argument('--verbose', '-v'    , action='count',help=verb_help)
    parser.add_argument('--print_summary'    , '-p', action='store_true', help=print_help)
    parser.add_argument('--version', action='version', version='%(prog)s ' + str(version))

    args = parser.parse_args()

    #args = parser.parse_args()
    # Constraints for the cnv files to search
    constraint_station = None
    print_summary = args.print_summary

    if(args.verbose == None):
        loglevel = logging.INFO
    elif(args.verbose == 1):
        loglevel = logging.WARNING
    elif(args.verbose == 2):
        loglevel = logging.INFO        
    elif(args.verbose > 2):
        loglevel = logging.DEBUG
    else:
        loglevel = logging.INFO


    logger.setLevel(loglevel)

    filename = args.filename
    if(filename != None):
        try:
            fi = open(filename,'w')
        except exception as e:
            logger.critical('Could not open file:' + filename)
            logger.critical('Exiting... ')
            exit()



    #
    # Distance option, and no station is defined
    #
    if(args.distance != None and args.station == None):
        FLAG_DIST=True
        londist  = float(args.distance[0])
        latdist  = float(args.distance[1])
        distdist = float(args.distance[2])
        # Creating a dictionary for the get_valid_files function
        constraint_station = [None,None,None] 
        constraint_station[0]  = londist
        constraint_station[1]  = latdist
        constraint_station[2]  = distdist                            
        logger.info('Will search for profiles within a ' + str(distdist) + ' m radius around station with longitude ' + str(londist) + ' and latitude ' + str(latdist))
        if(FLAG_PYPROJ == False):
            logger.critical('pyproj is not installed, cannot compute distance, exiting')
            exit()
    else:
        FLAG_DIST=False



    #
    # Read stations file
    #
    #print(args.list_stations,args.station)
    if(args.list_stations or args.station != None):
        FLAG_DIST=False
        if(args.station != None):
            logger.info('Looking for station with name:' + args.station[0])
            sname_tmp = args.station[0]
        else:
            sname_tmp = None
        stations_file = pkg_resources.resource_filename('pycnv', 'stations/iow_stations.yaml')
        print('Stations file:',stations_file)
        f_stations = open(stations_file)
        # use safe_load instead load
        stations_yaml = yaml.safe_load(f_stations)
        f_stations.close()

        for i,station in enumerate(stations_yaml['stations']):
            namestation = [station['name']]
            lonstation  = station['longitude']
            latstation  = station['latitude']
            try:
                anames = station['alternative_names']
            except:
                anames = []

                
            namestation.extend(anames)
            for sname in namestation:
                if(sname == sname_tmp):
                    FLAG_DIST=True
                    logger.info('Found station')
                    londist  = lonstation
                    latdist  = latstation
                    distdist = float(args.station[1])
                    # Creating a dictionary for the get_valid_files function
                    constraint_station = [None,None,None]
                    constraint_station[0]  = londist
                    constraint_station[1]  = latdist
                    constraint_station[2]  = distdist                    
                    logger.info('Will search for profiles within a ' + str(distdist) +\
                                 ' m radius around station ' + str(sname) + ' with longitude '\
                                  + str(londist) + ' and latitude ' + str(latdist))
                    if(FLAG_PYPROJ == False):
                        logger.critical('pyproj is not installed, cannot compute distance, exiting')
                        sys.exit(1)


            if(args.list_stations == True):
                print(namestation,lonstation,latstation)

        if(FLAG_DIST == False and sname_tmp != None):
            logger.critical('Could not find a station with name ' + args.station[0] +  ' in station file, exiting.')
            sys.exit(1)


    if(args.list_stations == True):
        sys.exit(0)


    if(args.data_folder != None):
        DATA_FOLDER = args.data_folder
        logger.info('Will search in folders:' + str(DATA_FOLDER))
    else:
        parser.print_help()
        logger.critical('Specify a data path to search for cnv files ... exiting')
        sys.exit(0)

    # Read in all potential cnv files, this "double" reading is (probably)
    # necessary for sorting them without saving all the data into RAM
    # TODO, if more speed is needed more data can be saved into cnv_data
    logger.info('Checking for double datasets')
    cnv_data = get_all_valid_files(DATA_FOLDER, loglevel = loglevel, station = constraint_station, save_summary = True)
    # Searching for files with the same origin (but probably different postprocessing of the seabird software)
    lon_d      = numpy.asarray(cnv_data['lon'])
    lat_d      = numpy.asarray(cnv_data['lat'])
    date_d     = numpy.asarray(cnv_data['dates'])
    checked_d  = numpy.zeros(len(lon_d),dtype=int)
    num_d      = numpy.zeros(len(lon_d),dtype=int)

                    
    for i in range(len(lon_d)):
        if(checked_d[i] == 0):
            checked_d[i]       = 1
            ind_date           = date_d[i] == date_d
            ind_lon            = lon_d[i] == lon_d
            ind_lat            = lat_d[i] == lat_d
            ind_all            = ind_date & ind_lon & ind_lat
            checked_d[ind_all] = 1
            num_d[ind_all]     = num_d.max() + 1

    
    file_names_save = cnv_data['files']
    # Save the valid files into the specified file or print it to console
    if True:
        num_wr = 0
        for nf,fname in enumerate(file_names_save):
            #FLAG_GOOD = save_file[ind]
            #if(FLAG_GOOD):
                #cnv = pycnv.pycnv(matches[ind],verbosity=logging.CRITICAL)
            if True:
                if(nf == 0):
                    cnv        = pycnv(fname,verbosity=logging.CRITICAL)
                    cnv_header = cnv.get_summary(header=True)                
                    #summary    = cnv.get_summary()
                    
                summary = cnv_data['summary'][nf]
                    
                # Adding information about double files
                sep = ','
                summary = '{:5d}'.format(nf) + sep + '{:5d}'.format(num_d[nf]) + sep + summary
                cnv_header = 'num file'+ sep + 'num double' + sep + cnv_header
                if(print_summary):
                    if(nf == 0):
                        print(cnv_header)
                        
                    print(summary)
                if(filename != None):
                    if(nf == 0):
                        fi.write(cnv_header)
                        fi.write('\n')
                    fi.write(summary)
                    fi.write('\n')
                    fi.flush()
                    num_wr +=1

    else:
        print('No files found ... ')

    nunique = 0
    if(len(num_d) > 0):
        nunique = num_d.max()
    logger.info('Read ' +str(len(file_names_save)) + ' files (' + str(nunique) + ' with unique datasets)')
    if(filename != None):
        logger.info('Wrote ' +str(num_wr) + ' datasets into file:' + filename)
        fi.close()



if __name__ == '__main__':
   main()    


    
